File: bin/docparser.rb
require 'pp'
require 'json'

$bad_input = false
def bad_input file, text
	$bad_input = true
	$stderr.puts "#{file}: unrecognized input: #{text}"
end

def parse_dir dirname
	Dir.entries(dirname).sort.map{|filename|
		if filename == '.' || filename == '..'
			nil
		else
			parse_any_path "#{dirname}/#{filename}"
		end
	}.compact.inject(:+)
end

def cleanup_class_name class_name
	class_name.sub(/OO\.ui\./, '').sub(/mixin\./, '')
end

def extract_default_from_description item
	m = item[:description].match(/\(default: (.+?)\)\s*?$/i)
	return if !m
	# modify `item` in-place
	item[:default] = m[1]
	item[:description] = m.pre_match
end

def parse_file filename
	if filename !~ /\.(php|js)$/
		return nil
	end
	filetype = filename[/\.(php|js)$/, 1].to_sym

	text = File.read filename, encoding: 'utf-8'

	# ewwww
	# some docblocks are missing and we really need them
	text = text.sub(/(?<!\*\/\n)^(class|trait)/, "/**\n*/\n\\1")

	# find all documentation blocks, together with the following line (unless it contains another docblock)
	docblocks = text.scan(/\/\*\*[\s\S]+?\*\/\n[ \t]*(?:(?=\/\*\*)|.*)/)

	current_class = nil
	output = []
	previous_item = {} # dummy

	docblocks.each{|d|
		kind = nil
		previous_item = data = {
			name: nil,
			description: '',
			parent: nil,
			mixes: [],
			methods: [],
			properties: [],
			events: [],
			params: [],
			config: [],
			visibility: :public,
			type: nil,
		}
		valid_for_all = %w[name description].map(&:to_sym)
		valid_per_kind = {
			class:    valid_for_all + %w[parent mixes methods properties events abstract].map(&:to_sym),
			method:   valid_for_all + %w[params config return visibility static].map(&:to_sym),
			property: valid_for_all + %w[type static].map(&:to_sym),
			event:    valid_for_all + %w[params].map(&:to_sym),
		}

		js_class_constructor = false
		js_class_constructor_desc = ''
		php_trait_constructor = false
		parsing_php_config_options = false
		ignore = false

		comment, code_line = d.split '*/'
		comment.split("\n").each{|comment_line|
			next if comment_line.strip == '/**'
			comment_line.sub!(/^[ \t]*\*[ \t]?/, '') # strip leading '*' and whitespace

			match_keyword = comment_line.match(/^@([\w-]+)(?:[ \t]+(.+))?/)
			match_php_config = comment_line.match(/^ *- (\S+) \&?(?:\.\.\.)?\$config(?:\['(\w+)'\])?( .+)?$/)
			if !match_keyword && parsing_php_config_options && match_php_config
				type, config, description = match_php_config.captures
				data[:config] << {name: config, type: cleanup_class_name(type), description: description || ''}
				previous_item = data[:config][-1]
				extract_default_from_description(previous_item)
				next
			end
			if !match_keyword
				# this is a continuation of previous item's description
				previous_item[:description] << comment_line + "\n"
				if filetype == :php
					extract_default_from_description(previous_item)
				end
				next
			end

			parsing_php_config_options = false
			keyword, content = match_keyword.captures

			# handle JS class/constructor conundrum
			if keyword == 'class' || keyword == 'constructor'
				js_class_constructor = true
			end

			case keyword
			when 'constructor'
				# handle JS class/constructor conundrum
				js_class_constructor_desc = data[:description]
				data[:description] = ''
				kind = :method
			when 'class'
				kind = :class
				data[:name] = cleanup_class_name(content.strip) if content && !content.strip.empty?
			when 'method'
				kind = :method
			when 'property', 'var'
				kind = :property
				m = content.match(/^\{?(.+?)\}?( .+)?$/)
				if !m
					bad_input filename, comment_line
					next
				end
				type, description = m.captures
				data[:type] = type
				data[:description] = description if description
			when 'event'
				kind = :event
				data[:name] = content.strip
			when 'extends'
				data[:parent] = cleanup_class_name(content.strip)
			when 'mixes'
				data[:mixes] << cleanup_class_name(content.strip)
			when 'param'
				case filetype
				when :js
					type, is_config, name, default, description =
						content.match(/^\{(?:\.\.\.)?(.+?)\} \[?(config\.)?([\w.$]+?)(?:=(.+?))?\]?( .+)?$/).captures
					# ignore the "Configuration options" parameter
					next if type == 'Object' && name == 'config'
					key = is_config ? :config : :params
					data[key] << {name: name, type: cleanup_class_name(type), description: description || '', default: default}
					previous_item = data[key][-1]
				when :php
					type, name, description =
						content.match(/^(\S+) \&?(?:\.\.\.)?\$(\w+)( .+)?$/).captures
					parsing_php_config_options = type == 'array' && name == 'config'
					# ignore the "Configuration options" parameter
					next if parsing_php_config_options
					data[:params] << {name: name, type: cleanup_class_name(type), description: description || ''}
					previous_item = data[:params][-1]
					extract_default_from_description(previous_item)
				end
			when 'return'
				case filetype
				when :js
					m = content.match(/^\{(.+?)\}( .+)?$/)
				when :php
					m = content.match(/^(\S+)( .+)?$/)
				end
				if !m
					bad_input filename, comment_line
					next
				end
				type, description = m.captures
				data[:return] = {type: cleanup_class_name(type), description: description || ''}
				previous_item = data[:return]
			when 'private'
				data[:visibility] = :private
			when 'protected'
				data[:visibility] = :protected
			when 'static'
				data[:static] = true
			when 'abstract'
				data[:abstract] = true
			when 'ignore', 'inheritdoc'
				ignore = true
			when 'inheritable', 'deprecated', 'namespace', 'throws',
				 'chainable', 'fires', 'memberof',
				 'see', 'uses', 'param-taint', 'suppress',
				 'phpcs',
				 /^phan-/
				# skip
			else
				bad_input filename, comment_line
				next
			end
		}

		next if ignore

		if code_line && code_line.strip != ''
			case filetype
			when :js
				m = code_line.match(/(?:(static|prototype|mixin)\.)?(\w+) =/)
				if !m
					bad_input filename, code_line.strip
					next
				end
				kind_, name = m.captures
				data[:static] = true if kind_ == 'static'
				kind = {'static' => :property, 'prototype' => :method}[ kind_.strip ] if kind_ && !kind
				data[:mixin] = true if kind_ == 'mixin'
				data[:name] ||= cleanup_class_name(name)
			when :php
				m = code_line.match(/
					\s*
					(?:(public|protected|private)\s)?
					(?:(static)\s)?(function\s|class\s|trait\s|\$)
					(\w+)
					(?:\sextends\s(\w+))?
				/x)
				if !m
					bad_input filename, code_line.strip
					next
				end
				visibility, static, kind_, name, parent = m.captures
				kind = {'$' => :property, 'function' => :method, 'class' => :class, 'trait' => :class}[ kind_.strip ]
				data[:visibility] = {'private' => :private, 'protected' => :protected, 'public' => :public}[ visibility ] || :public
				data[:mixin] = true if kind_.strip == 'trait'
				data[:static] = true if static
				data[:parent] = cleanup_class_name(parent) if parent
				data[:name] ||= cleanup_class_name(name)
				php_trait_constructor = true if kind == :method && data[:name] == 'initialize' + current_class[:name]
			end
		end

		# handle JS class/constructor conundrum
		if kind == :class || js_class_constructor
			if current_class
				output << current_class
			end
			current_class = data.select{|k, _v| valid_per_kind[:class].include? k }
			current_class[:description] = js_class_constructor_desc if js_class_constructor_desc != ''
			previous_item = current_class
		end

		# standardize
		# (also handle fake constructors for traits)
		if data[:name] == '__construct' || js_class_constructor || php_trait_constructor
			data[:name] = '#constructor'
		end

		# put into the current class
		if kind && kind != :class
			keys = {
				method: :methods,
				property: :properties,
				event: :events,
			}
			if current_class
				current_class[keys[kind]] << data.select{|k, _v| valid_per_kind[kind].include? k }
				previous_item = current_class[keys[kind]]
			end
		end
	}

	# this is evil, assumes we only have one class in a file, but we'd need a proper parser to do it better
	# note, we remove everything before the first "{\n" to make sure we don't parse `use` imports.
	if current_class && filetype == :php
		current_class[:mixes] +=
			text.gsub(/\A.+?\{\n/m, '').scan(/^[ \t]*use (\w+)(?: ?\{|;)/).flatten.map(&method(:cleanup_class_name))
	end

	output << current_class if current_class
	output
end

def parse_any_path path
	if File.directory? path
		result = parse_dir path
	else
		result = parse_file path
	end
	if $bad_input
		$stderr.puts 'Unrecognized inputs encountered, stopping.'
		exit 1
	end
	result
end

if __FILE__ == $PROGRAM_NAME
	if ARGV.empty? || ARGV == ['-h'] || ARGV == ['--help']
		$stderr.puts "usage: ruby #{$PROGRAM_NAME} <files...>"
		$stderr.puts "       ruby #{$PROGRAM_NAME} src > docs-js.json"
		$stderr.puts "       ruby #{$PROGRAM_NAME} php > docs-php.json"
	else
		out = JSON.pretty_generate ARGV.map{|a| parse_any_path a }.inject(:+)
		# ew
		out = out.gsub(/\{\s+\}/, '{}').gsub(/\[\s+\]/, '[]')
		puts out
	end
end

File: bin/docparser.js
'use strict';

const fs = require( 'fs' );
const path = require( 'path' );

let badInputSeen = false;

function badInput( file, text ) {
	badInputSeen = true;
	process.stderr.write( `${ file }: unrecognized input: ${ text }\n` );
}

function cleanupClassName( className ) {
	return className.replace( /^OO\.ui\./, '' ).replace( /^mixin\./, '' );
}

function extractDefaultFromDescription( item ) {
	const match = item.description && item.description.match( /\(default: (.+?)\)\s*?$/i );
	if ( !match ) {
		return;
	}
	// Modify `item` in-place
	item.default = match[ 1 ];
	item.description = item.description.slice( 0, match.index );
}

function parseDir( dirName ) {
	const entries = fs.readdirSync( dirName ).sort();
	const out = [];
	for ( const entry of entries ) {
		if ( entry === '.' || entry === '..' ) {
			continue;
		}
		out.push( ...parseAnyPath( path.join( dirName, entry ) ) );
	}
	return out;
}

function maybeInsertEmptyDocblock( text ) {
	// ewwww
	// some docblocks are missing and we really need them
	const match = text.match( /^(class|trait)\b/m );
	if ( !match ) {
		return text;
	}
	const index = match.index;
	const before = text.slice( 0, index );
	if ( before.endsWith( '*/\n' ) ) {
		return text;
	}
	return before + '/**\n*/\n' + text.slice( index );
}

function parseFile( fileName ) {
	if ( !/\.(php|js)$/.test( fileName ) ) {
		return [];
	}
	const fileType = fileName.endsWith( '.php' ) ? 'php' : 'js';
	let text = fs.readFileSync( fileName, { encoding: 'utf-8' } );

	text = maybeInsertEmptyDocblock( text );

	// Find all documentation blocks, together with the following line
	// (unless it contains another docblock)
	const docblockRegex = /\/\*\*[\s\S]+?\*\/\n[ \t]*(?:(?=\/\*\*)|.*)/g;
	const docblocks = text.match( docblockRegex ) || [];

	let currentClass = null;
	const output = [];
	let previousItem = { description: '' }; // Dummy

	for ( const block of docblocks ) {
		let kind = null;
		const data = {
			name: null,
			description: '',
			parent: null,
			mixes: [],
			methods: [],
			properties: [],
			events: [],
			params: [],
			config: [],
			visibility: 'public',
			type: null
		};

		const validForAll = new Set( [ 'name', 'description' ] );
		const validPerKind = {
			class: new Set( [ ...validForAll, 'parent', 'mixes', 'methods', 'properties', 'events', 'abstract' ] ),
			method: new Set( [ ...validForAll, 'params', 'config', 'return', 'visibility', 'static' ] ),
			property: new Set( [ ...validForAll, 'type', 'static' ] ),
			event: new Set( [ ...validForAll, 'params' ] )
		};

		let jsClassConstructor = false;
		let jsClassConstructorDesc = '';
		let phpTraitConstructor = false;
		let parsingPhpConfigOptions = false;
		let ignore = false;

		const splitIndex = block.indexOf( '*/' );
		const comment = splitIndex === -1 ? block : block.slice( 0, splitIndex );
		const codeLine = splitIndex === -1 ? '' : block.slice( splitIndex + 2 );

		for ( let commentLine of comment.split( '\n' ) ) {
			if ( commentLine.trim() === '/**' ) {
				continue;
			}
			// Strip leading '*' and whitespace
			commentLine = commentLine.replace( /^[ \t]*\*[ \t]?/, '' );

			const matchKeyword = commentLine.match( /^@([\w-]+)(?:[ \t]+(.+))?/ );
			const matchPhpConfig = commentLine.match( /^ *- (\S+) &?(?:\.\.\.)?\$config(?:\['(\w+)'\])?( .+)?$/ );

			if ( !matchKeyword && parsingPhpConfigOptions && matchPhpConfig ) {
				const [ , type, config, description ] = matchPhpConfig;
				data.config.push( {
					name: config,
					type: cleanupClassName( type ),
					description: description || ''
				} );
				previousItem = data.config[ data.config.length - 1 ];
				extractDefaultFromDescription( previousItem );
				continue;
			}

			if ( !matchKeyword ) {
				// This is a continuation of previous item's description
				previousItem.description = ( previousItem.description || '' ) + commentLine + '\n';
				if ( fileType === 'php' ) {
					extractDefaultFromDescription( previousItem );
				}
				continue;
			}

			parsingPhpConfigOptions = false;
			const keyword = matchKeyword[ 1 ];
			const content = matchKeyword[ 2 ];

			// Handle JS class/constructor conundrum
			if ( keyword === 'class' || keyword === 'constructor' ) {
				jsClassConstructor = true;
			}

			switch ( keyword ) {
				case 'constructor':
					// Handle JS class/constructor conundrum
					jsClassConstructorDesc = data.description;
					data.description = '';
					kind = 'method';
					break;
				case 'class':
					kind = 'class';
					if ( content && content.trim() !== '' ) {
						data.name = cleanupClassName( content.trim() );
					}
					break;
				case 'method':
					kind = 'method';
					break;
				case 'property':
				case 'var': {
					kind = 'property';
					const match = content && content.match( /^\{?(.+?)\}?( .+)?$/ );
					if ( !match ) {
						badInput( fileName, commentLine );
						break;
					}
					const [ , type, description ] = match;
					data.type = type;
					if ( description ) {
						data.description = description;
					}
					break;
				}
				case 'event':
					kind = 'event';
					data.name = content.trim();
					break;
				case 'extends':
					data.parent = cleanupClassName( content.trim() );
					break;
				case 'mixes':
					data.mixes.push( cleanupClassName( content.trim() ) );
					break;
				case 'param':
					if ( fileType === 'js' ) {
						const match = content.match( /^\{(?:\.\.\.)?(.+?)\} \[?(config\.)?([\w.$]+?)(?:=(.+?))?\]?( .+)?$/ );
						if ( !match ) {
							badInput( fileName, commentLine );
							break;
						}
						const [ , type, isConfig, name, def, description ] = match;
						// Ignore the "Configuration options" parameter
						if ( type === 'Object' && name === 'config' ) {
							break;
						}
						const key = isConfig ? 'config' : 'params';
						data[ key ].push( {
							name,
							type: cleanupClassName( type ),
							description: description || '',
							default: def
						} );
						previousItem = data[ key ][ data[ key ].length - 1 ];
					} else {
						// PHP
						const match = content.match( /^(\S+) &?(?:\.\.\.)?\$(\w+)( .+)?$/ );
						if ( !match ) {
							badInput( fileName, commentLine );
							break;
						}
						const [ , type, name, description ] = match;
						// Ignore the "Configuration options" parameter
						parsingPhpConfigOptions = type === 'array' && name === 'config';
						if ( parsingPhpConfigOptions ) {
							break;
						}
						data.params.push( {
							name,
							type: cleanupClassName( type ),
							description: description || ''
						} );
						previousItem = data.params[ data.params.length - 1 ];
						extractDefaultFromDescription( previousItem );
					}
					break;
				case 'return': {
					let match;
					if ( fileType === 'js' ) {
						match = content && content.match( /^\{(.+?)\}( .+)?$/ );
					} else {
						match = content && content.match( /^(\S+)( .+)?$/ );
					}
					if ( !match ) {
						badInput( fileName, commentLine );
						break;
					}
					const [ , type, description ] = match;
					data.return = {
						type: cleanupClassName( type ),
						description: description || ''
					};
					previousItem = data.return;
					break;
				}
				case 'private':
					data.visibility = 'private';
					break;
				case 'protected':
					data.visibility = 'protected';
					break;
				case 'static':
					data.static = true;
					break;
				case 'abstract':
					data.abstract = true;
					break;
				case 'ignore':
				case 'inheritdoc':
					ignore = true;
					break;
				case 'inheritable':
				case 'deprecated':
				case 'namespace':
				case 'throws':
				case 'chainable':
				case 'fires':
				case 'memberof':
				case 'see':
				case 'uses':
				case 'param-taint':
				case 'suppress':
				case 'phpcs':
					break;
				default:
					if ( /^phan-/.test( keyword ) ) {
						break;
					}
					badInput( fileName, commentLine );
			}
		}

		if ( ignore ) {
			continue;
		}

		if ( codeLine && codeLine.trim() !== '' ) {
			if ( fileType === 'js' ) {
				const match = codeLine.match( /(?:(static|prototype|mixin)\.)?(\w+) =/ );
				if ( !match ) {
					badInput( fileName, codeLine.trim() );
					continue;
				}
				const [ , kindHint, name ] = match;
				if ( kindHint === 'static' ) {
					data.static = true;
				}
				if ( kindHint && !kind ) {
					kind = { static: 'property', prototype: 'method' }[ kindHint ];
				}
				if ( kindHint === 'mixin' ) {
					data.mixin = true;
				}
				data.name = data.name || cleanupClassName( name );
			} else {
				const match = codeLine.match( /\s*(?:(public|protected|private)\s)?(?:(static)\s)?(function\s|class\s|trait\s|\$)(\w+)(?:\sextends\s(\w+))?/ );
				if ( !match ) {
					badInput( fileName, codeLine.trim() );
					continue;
				}
				const [ , visibility, staticKeyword, kindToken, name, parent ] = match;
				const kindTokenTrim = kindToken.trim();
				kind = { $: 'property', function: 'method', class: 'class', trait: 'class' }[ kindTokenTrim ];
				data.visibility = { private: 'private', protected: 'protected', public: 'public' }[ visibility ] || 'public';
				if ( kindTokenTrim === 'trait' ) {
					data.mixin = true;
				}
				if ( staticKeyword ) {
					data.static = true;
				}
				if ( parent ) {
					data.parent = cleanupClassName( parent );
				}
				data.name = data.name || cleanupClassName( name );
				if ( kind === 'method' && data.name === 'initialize' + ( currentClass ? currentClass.name : '' ) ) {
					phpTraitConstructor = true;
				}
			}
		}

		// Handle JS class/constructor conundrum
		if ( kind === 'class' || jsClassConstructor ) {
			if ( currentClass ) {
				output.push( currentClass );
			}
			currentClass = {};
			for ( const [ key, value ] of Object.entries( data ) ) {
				if ( validPerKind.class.has( key ) ) {
					currentClass[ key ] = value;
				}
			}
			if ( jsClassConstructorDesc !== '' ) {
				currentClass.description = jsClassConstructorDesc;
			}
			previousItem = currentClass;
		}

		// Standardize
		// (also handle fake constructors for traits)
		if ( data.name === '__construct' || jsClassConstructor || phpTraitConstructor ) {
			data.name = '#constructor';
		}

		// Put into the current class
		if ( kind && kind !== 'class' ) {
			const destKey = { method: 'methods', property: 'properties', event: 'events' }[ kind ];
			if ( currentClass && destKey ) {
				const filtered = {};
				for ( const [ key, value ] of Object.entries( data ) ) {
					if ( validPerKind[ kind ].has( key ) ) {
						filtered[ key ] = value;
					}
				}
				currentClass[ destKey ].push( filtered );
				previousItem = filtered;
			}
		}
	}

	// This is evil, assumes we only have one class in a file, but we'd need a
	// proper parser to do it better.
	// Note, we remove everything before the first "{\n" to make sure we
	// don't parse `use` imports.
	if ( currentClass && fileType === 'php' ) {
		let stripped = text;
		stripped = stripped.replace( /^[\s\S]+?\{\n/m, '' );
		const useRegex = /^[ \t]*use (\w+)(?: ?\{|;)/gm;
		let match;
		while ( ( match = useRegex.exec( stripped ) ) !== null ) {
			currentClass.mixes.push( cleanupClassName( match[ 1 ] ) );
		}
		// Unique
		currentClass.mixes = Array.from( new Set( currentClass.mixes ) );
	}

	if ( currentClass ) {
		output.push( currentClass );
	}

	return output;
}

function parseAnyPath( inputPath ) {
	let result;
	if ( fs.statSync( inputPath ).isDirectory() ) {
		result = parseDir( inputPath );
	} else {
		result = parseFile( inputPath );
	}
	if ( badInputSeen ) {
		process.stderr.write( 'Unrecognized inputs encountered, stopping.\n' );
		process.exit( 1 );
	}
	return result;
}

module.exports = {
	parseAnyPath
};
